src/tally/KDEMeshTally.cpp

```diff
@@ -0,0 +1,586 @@
+      weight *= event.track_length;
+      subtrack_points = choose_points(num_subtracks, event);
+    }
+  } else if (event.type == TallyEvent::COLLISION && estimator == COLLISION) {
+    // divide weight by cross section and update optimal bandwidth
+    weight /= event.total_cross_section;
+    update_variance(event.position);
+  } else { // NONE, return from this method
+    return;
+  }
+
+  unsigned int ebin;
+  if (!get_energy_bin(event.particle_energy, ebin)) {
+    return;
+  }
+
+  // update the neighborhood region and find all of the calculations points
+  region->update_neighborhood(event, bandwidth);
+  const std::set<moab::EntityHandle>& calculation_points = region->get_points();
+
+  // iterate through calculation points and compute their final scores
+  std::set<moab::EntityHandle>::iterator i;
+  CalculationPoint X;
+
+  for (i = calculation_points.begin(); i != calculation_points.end(); ++i) {
+    // get coordinates of this point
+    moab::EntityHandle point = *i;
+    moab::ErrorCode rval = mbi->get_coords(&point, 1, X.coords);
+
+    assert(rval == moab::MB_SUCCESS);
+
+    // get tag data for this point if user requested boundary correction
+    if (use_boundary_correction) {
+      rval = mbi->tag_get_data(boundary_tag, &point, 1, X.boundary_data);
+      assert(rval == moab::MB_SUCCESS);
+
+      rval = mbi->tag_get_data(distance_tag, &point, 1, X.distance_data);
+      assert(rval == moab::MB_SUCCESS);
+    }
+
+    // compute the final contribution to the tally for this point
+    double score = 0.0;
+
+    if (estimator == INTEGRAL_TRACK) {
+      score = integral_track_score(X, event);
+    } else if (estimator == SUB_TRACK) {
+      score = subtrack_score(X, subtrack_points);
+    } else { // estimator == COLLISION
+      score = evaluate_kernel(X, event.position);
+    }
+
+    add_score_to_mesh_tally(point, weight, score,  ebin);
+  }  // end calculation_points iteration
+}
+//---------------------------------------------------------------------------//
+void KDEMeshTally::write_data(double num_histories) {
+  // display the optimal bandwidth if it was computed
+  if (estimator == COLLISION) {
+    std::cout << std::endl << "optimal bandwidth for " << num_collisions
+              << " collisions is: " << get_optimal_bandwidth() << std::endl;
+  }
+
+  // tag tally and relative error results to the mesh for each tally point
+  moab::ErrorCode rval = moab::MB_SUCCESS;
+  if (rval != moab::MB_SUCCESS) {
+    std::cout << "This can't fail" << std::endl;
+    exit(1);
+  }
+  moab::Range::iterator i;
+
+  for (i = tally_points.begin(); i != tally_points.end(); ++i) {
+    moab::EntityHandle point = *i;
+    unsigned int point_index = get_entity_index(point);
+
+    unsigned int num_ebins = data->get_num_energy_bins();
+    // if there is a total, dont do anything with it
+    if (data->has_total_energy_bin())
+      num_ebins--;
+
+    double tally_vect[num_ebins];
+    double error_vect[num_ebins];
```

Comment:
Variable length arrays are not supported in C++, though since they are in newer versions of C many C++ compilers support them. I'll suggest the changes required to use vectors instead here, but it's not necessarily required to use them. For these two lines, change them to:
```c++
  std::vector<double> tally_vect(num_ebins);
  std::vector<double> error_vect(num_ebins);
```
One more set of changes (two lines) below.